File: src/lexer/token_types.py
"""
Token type definitions for the Tombo lexer.
"""
from enum import Enum, auto


class LexerError(Exception):
    """Raised when lexer encounters an error."""

    def __init__(self, message: str, line: int, column: int):
        self.message = message
        self.line = line
        self.column = column
        super().__init__(f"{message} at line {line}, column {column}")


class TokenType(Enum):
    """All token types used by the Tombo lexer."""

    # Literals
    NUMBER = auto()
    STRING = auto()
    IDENTIFIER = auto()
    BOOLEAN = auto()

    # Keywords
    LET = auto()
    CHANGE = auto()
    TO = auto()
    DEF = auto()
    DEFI = auto()
    IF = auto()
    ELSE = auto()
    ELIF = auto()
    FOR = auto()
    WHILE = auto()
    IN = auto()
    END = auto()
    RETURN = auto()
    MATCH = auto()
    WHEN = auto()
    CLASS = auto()
    SELF = auto()
    IMPORT = auto()
    USE = auto()
    TRY = auto()
    CATCH = auto()
    FINALLY = auto()
    ASYNC = auto()
    AWAIT = auto()
    BREAK = auto()
    CONTINUE = auto()
    PASS = auto()
    THEN = auto()

    # Operators
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    POWER = auto()
    DOT = auto()
    COMMA = auto()
    COLON = auto()
    ARROW = auto()  # =>
    EQ = auto()  # =
    EQEQ = auto()  # ==
    NEQ = auto()  # !=
    LT = auto()  # <
    GT = auto()  # >
    LTE = auto()  # <=
    GTE = auto()  # >=
    AND = auto()  # and
    OR = auto()  # or
    NOT = auto()  # not
    AMPERSAND = auto()  # &
    PIPE = auto()  # |
    CARET = auto()  # ^
    TILDE = auto()  # ~
    LSHIFT = auto()  # <<
    RSHIFT = auto()  # >>
    PLUSEQ = auto()  # +=
    MINUSEQ = auto()  # -=
    STAREQ = auto()  # *=
    SLASHEQ = auto()  # /=

    # Delimiters
    LPAREN = auto()  # (
    RPAREN = auto()  # )
    LBRACKET = auto()  # [
    RBRACKET = auto()  # ]
    LBRACE = auto()  # {
    RBRACE = auto()  # }

    # Special
    INDENT = auto()
    DEDENT = auto()
    NEWLINE = auto()
    EOF = auto()

    # Comments (not normally tokens, but tracked)
    COMMENT = auto()


class Token:
    """Represents a single token in the source code."""

    def __init__(
        self, type_: TokenType, value: str, line: int, column: int, raw: str = None
    ):
        self.type = type_
        self.value = value
        self.line = line
        self.column = column
        self.raw = raw or value  # Original text from source

    def __repr__(self):
        return f"Token({self.type.name}, {repr(self.value)}, {self.line}:{self.column})"

    def __eq__(self, other):
        if not isinstance(other, Token):
            return False
        return (
            self.type == other.type
            and self.value == other.value
            and self.line == other.line
            and self.column == other.column
        )

File: tombo-rust/src/lexer/token_types.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Literals
    Integer(i64),
    Float(f64),
    String(String),
    Boolean(bool),

    // Keywords
    Let,
    Change,
    To,
    Def,
    If,
    Else,
    Elif,
    For,
    While,
    In,
    End,
    Return,
    Match,
    When,
    Class,
    Self_,
    Import,
    Use,
    Try,
    Catch,
    Finally,
    Async,
    Await,
    Break,
    Continue,
    Pass,
    Then,
    And,
    Or,
    Not,

    // Operators
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    DoubleSlash,
    DoubleStar,
    Equal,
    DoubleEqual,
    NotEqual,
    Less,
    Greater,
    LessEqual,
    GreaterEqual,
    Ampersand,
    Pipe,
    Caret,
    Tilde,
    DoubleAmpersand,
    DoublePipe,
    PlusEqual,
    MinusEqual,
    StarEqual,
    SlashEqual,
    DoubleSlashEqual,
    PercentEqual,
    DoubleStarEqual,
    AmpersandEqual,
    PipeEqual,
    CaretEqual,
    LeftShift,
    RightShift,
    LeftShiftEqual,
    RightShiftEqual,

    // Delimiters
    LeftParen,
    RightParen,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Comma,
    Colon,
    Semicolon,
    Dot,
    Arrow,
    FatArrow,
    Question,

    // Special
    Indent,
    Dedent,
    Newline,
    Eof,

    // Identifiers
    Identifier(String),
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(token_type: TokenType, line: usize, column: usize) -> Self {
        Token {
            token_type,
            line,
            column,
        }
    }
}

#[derive(Debug)]
pub struct LexerError {
    pub message: String,
    pub line: usize,
    pub column: usize,
}

impl fmt::Display for LexerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "Lexer error at {}:{}: {}",
            self.line, self.column, self.message
        )
    }
}

impl std::error::Error for LexerError {}
